File: projs/cpp/bb2.cpp
#include <iostream>
#include <vector>
#include <string>

int main() {
    std::vector<std::string> testStrings = {
        "{([])}{}", "{([)]}", "{{[[(())]]}}", "([{}])", "([)]", "", "{[()()]}", "{[(])}",
        "{{{{}}", "[({})]", "{[}]", "()[]{}", "((()))", "((())", "{{}}[()]{{}}", "{[({[({[()]})]})]}",
        "{[({[({[()}])]}])}", "{[]}", "{[({})]}", "{[(])}"
    };

    std::vector<bool> expectedResults = {
        true, false, true, true, false, true, true, false,
        false, true, false, true, true, false, true, true,
        false, true, true, false
    };

    bool allTestsPassed = true;

    for (int i = 0; i < testStrings.size(); ++i) {
        bool result = is_balanced(testStrings[i]);  // Calling the is_balanced function

        if (result != expectedResults[i]) {
            std::cout << "Test failed for string: \"" << testStrings[i] << "\". Expected: ";
            if (expectedResults[i]) {
                std::cout << "balanced";
            } else {
                std::cout << "not balanced";
            }
            std::cout << ", but got: ";
            if (result) {
                std::cout << "balanced";
            } else {
                std::cout << "not balanced";
            }
            std::cout << std::endl;
            allTestsPassed = false;
        }
    }

    if (allTestsPassed) {
        std::cout << "All tests passed." << std::endl;
    }

    return 0;
}

File: projs/py/bb2.py
def main():
    test_strings = [
        "{([])}{}", "{([)]}", "{{[[(())]]}}", "([{}])", "([)]", "", "{[()()]}", "{[(])}",
        "{{{{}}", "[({})]", "{[}]", "()[]{}", "((()))", "((())", "{{}}[()]{{}}", "{[({[({[()]})]})]}",
        "{[({[({[()}])]}])}", "{[]}", "{[({})]}", "{[(])}"
    ]

    expected_results = [
        True, False, True, True, False, True, True, False,
        False, True, False, True, True, False, True, True,
        False, True, True, False
    ]

    all_tests_passed = True

    for i in range(len(test_strings)):
        result = is_balanced(test_strings[i])

        # Debug output to check each test result
        print(f'Testing "{test_strings[i]}": ', end="")
        print("balanced" if result else "not balanced", end=" - Expected: ")
        print("balanced" if expected_results[i] else "not balanced")

        # Check if the result matches the expected result
        if result != expected_results[i]:
            print(f'Test failed for "{test_strings[i]}". Expected: '
                  f'{"balanced" if expected_results[i] else "not balanced"}, but got: '
                  f'{"balanced" if result else "not balanced"}')
            all_tests_passed = False

    if all_tests_passed:
        print("All tests passed.")

if __name__ == "__main__":
    main()
